Convert to emit 'loose methods' with explicit 'self' pointers

gezalore opened the pull request:
This is a bigish change. It implements #2988  (which itself is on the way to #2958 and #2140). There are some minor differences from what was discussed in #2988 but the result is hopefully more practicable.

It is still WIP for the reasons raised below, on which I would like some advice, but it is mostly there. Regressions, ext tests pass and https://github.com/lowRISC/opentitan builds.

I tried to keep the commits self consistent and tests shold pass at all of them, in case this helps review.

I will try to summarize the effective change in the output code at the end of this patch queue.

Internally, there are 2 ways to reference variables/cells (as in V3Descope). Relative addressing via `this`, or absolute addressing via `vlSyms`. This means any internal function may require access to these 2 values.

In this PR, all internal functions that are under an `AstNodeModule` (except under `AstClass`) have been changed to receive an explicit `self` pointer as a single argument, and then converted to what I call 'loose methods'. Loose methods are semantically methods in the sense that they associate with a particular class, but they are declared as global functions, so they don't need to be declared in the header (which removes the dependency causing recompilatoin). This means the standard signature changed from:
```
<ModuleClass>::<funcname>(Syms* vlSyms);
```
to:
```
<ModuleClass>__<funcname>(<ModuleClass>* self);
```

In all modules bar AstClass, `vlSyms` is available via `self->vlSyms`. There are hints towards aliasing issues in V3Descope, but these seem to relate to vintage compliers, the workaround for which (--no-relative-cfuncs) is already deprecated. At a later date we should change things to use relative addressing wherever possible for simplicity. The issue seems to be the potential aliasing between `self` and `vlSyms->SOMETHING_WIHT_TYPE_OF_SELF`. I think if this ever comes out again, we should just teach Verilator to emmit better code that work with mixed references by the judiscious use or `restrict`, unless you think we need to worry about vintage compilers. Comments or concerns on this are welcome.

There is no change to methdos under an `AstClass`, they are emitted as methods before, and `vlSyms` is passed as a parameter as before.

To keep Verilator internals simple, relative references are always emitted as an acces via explicit poitner access by emitting `this->...`, and aboslute references are emitted as `vlSyms->...`. For normal methods, this just works because `this` is available, and `vlSyms` either comes from a function argument (for `AstClass` members) or a class member (other `AstNodeModule` members). For loose methods, we emit a standard prologue and epliloge of the form:
```
#define this self
#define vlSyms self->vlSyms
  <body>
#undef this
#undef vlSyms
```

Once we introduced loose methods, we now no longer need to declare them in the module header, so we just declare them lazily as needed.

This all seems to work. I run some basic benchmarking on EH1, and it looks ok:
```
Platform:
- Intel(R) Core(TM) i7-5930K
- Turbo boost disabled
- CPU frequency pinned to 3.50GHz
- Everything run single threaded on core 0
- Tracing to ramdisk
- System otherwise idle
- Ubuntu 20.04.2 LTS
- Compilers installed via apt

################################################################################
# EH1 - no tracing

make -f $RV_ROOT/tools/Makefile \
  VERILATOR=$VERILATOR \
  CONF_PARAMS="-iccm_enable=1 -fpga_optimize" \
  target=high_perf \
  TEST=cmark_iccm \
  verilator-build program.hex

numactl -C 0 perf stat -r 10 ./obj_dir/Vtb_top

+--------------+----------------+----------------+
| compiler     | loose methods  | 1f331bd9       |
+==============+================+================+
| gcc    9.3.0 | 18.30 +- 0.004 | 18.33 +- 0.004 |
| clang 10.0.0 | 19.44 +- 0.006 | 19.49 +- 0.005 |

####################
# EH1 - tracing

NOTE: testbench/test_tb_top.cpp modified to write sim.vcd to ramdisk

make -f $RV_ROOT/tools/Makefile \
  VERILATOR=$VERILATOR \
  CONF_PARAMS="-iccm_enable=1 -fpga_optimize" \
  target=high_perf \
  TEST=cmark_iccm \
  debug=1 \
  verilator-build program.hex

numactl -C 0 perf stat --pre "rm -f /ramdisk/sim.vcd" -r 10 ./obj_dir/Vtb_top +dumpon

+--------------+----------------+----------------+
| compiler     | loose methods  | 1f331bd9       |
+==============+================+================+
| gcc    9.3.0 | 55.81 +- 0.004 | 55.75 +- 0.008 |
| clang 10.0.0 | 42.62 +- 0.005 | 42.69 +- 0.014 |
```

The reason the PR is still WIP:
- I actually need to test with the hashing base code splitting I have been working on that this actually works, but it addresses the issues I saw previously. I will do this next.
- You might want to run some more performance benchmarks. Sadly I don't have access to large designs to do much more testing. I requested some suggestions from the opentitan team for using that project as a benchmark (looking at you @rswarbrick) as it seems to be of sufficient size.
- In all of this, there is a need to figure out which `AstNodeModule` a particular `AstCFunc` is under (for cross module references), I have been using the following gross hack to do this, which is not ideal:
https://github.com/gezalore/verilator/blob/a31b3fbed6ae904e3b2b57ebbc5899a0bb39af75/src/V3EmitC.cpp#L391-L393
Note in later patches there will be a need finding the module of any `AstVar` as well (again for cross module references to static constants for example), in which case this hack is not only gross but is a performance problem due to `backp` going backwards on a long list before moving up the tree at the head of the list. Having this mechanism also allows us to get rid of the `classPrefix` member from `AstVarRef`/`AstNodeCCall` (currently used to access superclass methods/static variable, which would be nice. Do you have a suggestion on what would be the best way of doing this? I worry having  a `m_modp` inside an `AstCFunc`/`AstVar` that points to the module up the tree would complicate internals having to keep it in sync when movingthings around. Also `m_scopep` seems not available all the way to Emit.

EDIT: fixed typos

rswarbrick commented:
> I requested some suggestions from the opentitan team for using that project as a benchmark (looking at you @rswarbrick) as it seems to be of sufficient size.

Sorry for the slow reply! @imphil has helped me find such a command. The trick is to do a full chip simulation with a binary (running on the chip's CPU) that doesn't take too many cycles. This is normally wrapped up in a load of pytest magic, but here's a "direct" way to run stuff:
1. Clone OpenTitan and install the [build requirements](https://docs.opentitan.org/doc/ug/install_instructions/index.html). These include a RISC-V toolchain.
2. Build binaries to run on the CPU. The easiest way to do this is to run `./meson_init.sh` and then `ninja -C build-out all` at top-level (should take a minute or so).
3. Build the chip-level simulation with Verilator, as described [here](https://docs.opentitan.org/doc/ug/getting_started_verilator/). In practice, this boils down to `fusesoc --cores-root . run --flag=fileset_top --target=sim --setup --build lowrisc:systems:chip_earlgrey_verilator`.
4. Run the test...

For step (4), here's a test that takes roughly 30 seconds on a fast laptop:

`build/lowrisc_systems_chip_earlgrey_verilator_0.1/sim-verilator/Vchip_earlgrey_verilator --meminit=rom,build-bin/sw/device/boot_rom/boot_rom_sim_verilator.scr.40.vmem  --meminit=flash,build-bin/sw/device/tests/dif_uart_smoketest_sim_verilator.elf --meminit=otp,build-bin/sw/device/otp_img/otp_img_sim_verilator.vmem +UARTDPI_LOG_uart0=-`

(yes, the command line is crazy, but that's probably more helpful than using the pytest wrapper).

If you want something a bit longer, switch to `--meminit=flash,build-bin/sw/device/tests/otbn_ecdsa_p256_test_sim_verilator.elf`. This should take something like 5 minutes.

If you don't want to install the RISC-V toolchain, you can skip point (2) above and download the binaries from a recent CI run. The HW/SW interfaces are reasonably stable, so this *should* work, but I'd probably recommend building from scratch to make sure nothing gets out of sync.

gezalore commented:
An importatn point to add is that having `this` available in all methods (including loose ones) makes `$c` statements a lot more manageable to write without worrying about how Verilator will emit the containing function. You still need to write `this->` explicitly, but that is usually the right reference. See the updated tests for why this matters.

wsnyder commented:
Meta feedback (ignoring code for moment).

Seems great stuff!

Perhaps it's time to make a patch to rip out --no-relative-cfuncs before doing this one?  You're right this was getting around the compiler being stupid about pointer disambiguation, so is old stuff. Up to you.

I'd like to see no #defines, as they are confusing and it might mess up some lambdas like those used for "with" statements.  Instead we should have the code emit using e.g. "self" directly.  You probably could hack this by just having emit replace "this" with "self" when emitting a static.

Likewise no define of vlSyms.  Perhaps we should load vlSyms into a const variable up front of each function and not recompute self->vlSyms.  I generally prefer to err on the side of the compiler being stupid about dereferences and avoiding them when possible.

Performance looks great.

As you noted following backp is not nice.  Can they just be remembered when visiting the appropriate parent node in the parent's visitors?  Or worse case add a new pass that walks the tree and sets up user#p's to point to the appropriate parents?

wsnyder commented on test_regress/t/t_dpi_arg_inout_type__Dpi.out:
```diff
@@ -10,8 +10,8 @@
 #endif
- 
- 
+
+
```
Thanks for fixing the extra spaces - can you please throw that into a separate pull request and merge it into master - that'll cut down the number of files changing. Thanks.

gezalore replied:
Done

wsnyder commented on src/V3File.h:
```diff
@@ -165,6 +165,9 @@ class V3OutFormatter VL_NOT_FINAL {
     void blockDec() {
         if (!m_parenVec.empty()) m_parenVec.pop();
     }
+    void ensureNewLine() {
+        if (!m_nobreak) puts("\n");
+    }
```
Likewise you can please merge this into master (along with DPI spacing).

gezalore replied:
I think this is only necessary becasuse #define needs to go on a new line, so if we are not using them, we can loose this, though might be handy to make the output prettier. In any case, this is not used outside of this patch, so probably should not be separate (as it would then be unused). Note fixing the trailing whitespace was necessary to make that condition work I will merge that separately as requested.

wsnyder commented:
I see now what you mean by using $c("this->").  I guess the good news is with your define these $c's should run also on old versions.  $c("self->") would not.   $c isn't used a lot so I can probably get over it, but I dislike breaking old code like this.  What if we didn't use loose methods with any $c present?

gezalore commented:
> Perhaps it's time to make a patch to rip out --no-relative-cfuncs before doing this one?

Would love to, but the manual says:
https://github.com/verilator/verilator/blame/master/docs/guide/deprecations.rst#L16
It's not quite July yet, and taking out --no-relative-cfuncs  would not make any real difference to this patch so I'm happy to honour the scheduled deprecation and do a few releases from now.

> I guess the good news is with your define these $c's should run also on old versions. $c("self->") would not. 

Well, depends on what's in the `$c`, because for example `vlTOPp` which used to be in the standard prologue is now completely gone, so still a bit hit and miss...

I will have to look into the with statement and lambdas (I haven't noticed them being used before), but I don't mind de-macroing. You are right it can also just do `Syms* const __restrict vlSyms = self->vlSyms;` which indeed should help a lousy compiler.

> Can they just be remembered when visiting the appropriate parent node in the parent's visitors?

Not quite, because the callee might be in a different class than the caller, so it would be ordering dependent which we don't know up front. Stashing in user#p at the beginning of in `V3Emit::emit*` is probably not too bad so I will go with that. It has to be done twice, once for model, and once for tracing (otherwise the AstUser#InUse needs to go in an awkward place), but at some point I can just make the tracing less special and unify the two...

> $c isn't used a lot so I can probably get over it, but I dislike breaking old code like this. What if we didn't use loose methods with any $c present?

We could certainly do that, though I foresee a future when people will ask for the same loosening of class methods (we can just emit a manual vtable for virtuals), at which point proper methods might just be gone. It would be a pity to keep the complexity if not justified. I think it's a matter of when this needs to be broken rather than if as the context of the $c is entirely Verilator generated. 

Just curiosity: are there killer use cases for $c? All I ever used it for was to write Verilator test.

Note: I will also revert the pointless string splitting, my formatter seems to have gone a bit off the rails...

wsnyder commented:
I don't think $c is used much hence I do think we can break them if we have no other clean option - $c used to be used a lot before DPI import/exports were a thing.   I retract my suggestion to have $c turn off loose-ing.

gezalore commented:
> I'd like to see no #defines, as they are confusing and it might mess up some lambdas like those used for "with" statements. Instead we should have the code emit using e.g. "self" directly.

Actually the world is a bit more broken without 'this' beign available everywher, e.g. in t_func_rand there is a public function `QxRand32`, which is both a method on the class (because it's public) so 'this' is available, but is also inlined into a __sequent function, where without the define, only 'self' is available. This means that there is no way to use relative references in `QxRand32` in the $c present in that unction. Note in this specific case, I can still fix it by using `vlSymsp->TOPp`, but this does show the issue of various contexts where a single $c might be inlined into might differ... Any preference on how to proceed @wsnyder ? As I said, this one test I can just fix by using an absolute reference...

wsnyder commented:
Maybe use "self" in generated code, and add the define only if a $c is used in that function?  That way the $c is clear, as is the define only being present when needed.

gezalore commented:
Hmm, wouldn't that have the same potential issues as `with` and lambadas (I'm not sure what those issues are by the way)? 
At this point it might be best to just do s/this->/self->/ on $c contents when necessary and document this in the manual.

Note: I tried to add the #define/#undef only around the AstText nodes under a $c, but that does not work because sometimes $c  apperas within a macro argument, which the C compilers seem no to like, e.g. this does not compile, because VL_UNLIKELY is a macro:
```
if (VL_UNLIKELY((1U !=`
    // $c function at t/t_func_public.v:112:10
#define this self
this->publicInstNum()
#undef this
    ))) {
        VL_STOP_MT("t/t_func_public.v", 112, "");
}
```

wsnyder commented:
Or just text replace "this" (as a word) with "self" when spitting out the $c?

gezalore commented:
Ok, I removed all #defines (went with the word replace 'this' -> 'self' inside $c) and removed the hack to get to the parent module. I still need to do benchmarking and check the hash splits, which I will try to do tomorrow.

gezalore commented:
Notes on ccache efficiency:

As I mentioned before, my initial experiment on improvin the caching efficiency via the hash based splitting described in #2958 was useless because the function names changed in the global header. I am happy to report that after rebasing that work on this PR, that has now significantly improved. 

Measurement methodology:

When I prototyped the hash splits, I used a large design where everything was inlined (as every module was a singleton). I have since lost access to that design, so I am now using OpenTitan instead. OpenTitan is a bit different in that it has a lot of non-inlined modules, which presents additional issues for generating ccache friendly code while keeping builds fast (described below). To avoid this, I just added --flatten to Verilator to inline everything. After this, I ran 2 pairs of  builds, each of the pairs consist of:
- First, a build with an empty ccache
- Then, a build rebuilding the same (using the now pupulated ccache), but with adding 6 assignments to the CPU ALU that are semantically a no-op but Verilator does not understand that and emits new code

Results:

The first pair of builds was with this PR in pace, but with no hash based splitting. The ccache hit rate reported on the re-build using the tooling in #3011 is:
```
cache hit (direct) | 16 (4.98%)
cache miss         | 305 (95.02%)
```
The hits are basically the first 5 verilator output files which contain the statements before the injected assignments plus library files that don't change.

The second pair of builds was the same, but with the hash based splitting prototype in place, on top of this PR. I had to hand tune it a bit, and it only did 6 splits due to hashes:
```
cache hit (direct) | 553 (78.22%)
cache miss         | 154 (21.78%)
```

That looks like a win, though note there are a lot more files in the hash based splitting build. This is because in that mode, every function is emitted into a separate file (as if --output-split was set to 1). This itself can slow down the build considerably (and is the reason why I needed to add the --flatten, otherwise we would have ~7k files on OpenTitan) and will have to be addressed at a later state.

wsnyder commented:
Yes, good results - think this is looking good!

gezalore commented:
I have run  a bit more profiling with OpenTitan:
```
sudo numactl -C 0 chrt -f 99 perf stat -r 5

gcc - master:
      76,823.54 msec task-clock                #    1.000 CPUs utilized            ( +-  0.03% )
            100      context-switches          #    0.001 K/sec                    ( +- 35.69% )
              0      cpu-migrations            #    0.000 K/sec
          1,686      page-faults               #    0.022 K/sec
268,725,700,523      cycles                    #    3.498 GHz                      ( +-  0.03% )
254,769,408,085      instructions              #    0.95  insn per cycle           ( +-  0.00% )
 12,430,573,475      branches                  #  161.807 M/sec                    ( +-  0.00% )
    283,008,116      branch-misses             #    2.28% of all branches          ( +-  0.48% )

        76.8273 +- 0.0265 seconds time elapsed  ( +-  0.03% )


gcc - loose methods:

      79,952.02 msec task-clock                #    1.000 CPUs utilized            ( +-  0.05% )
             72      context-switches          #    0.001 K/sec                    ( +- 28.95% )
              0      cpu-migrations            #    0.000 K/sec
          1,725      page-faults               #    0.022 K/sec
279,670,960,510      cycles                    #    3.498 GHz                      ( +-  0.05% )
256,617,459,957      instructions              #    0.92  insn per cycle           ( +-  0.00% )
 12,501,713,177      branches                  #  156.365 M/sec                    ( +-  0.00% )
    304,927,011      branch-misses             #    2.44% of all branches          ( +-  0.42% )

        79.9549 +- 0.0397 seconds time elapsed  ( +-  0.05% )
```

That is a bit concerning as the GCC build is about 4% slower with loose methods, which is not ideal. I will investigate this more before we consider merging this.

gezalore commented:
I have tracked down and fixed the performance regression (see 20412ec). With that the performance is basically identical to maser on OpenTitan as well (thanks @rswarbrick and @imphil for the pointers):

```
gcc -- loose methods fixed:

      76,817.65 msec task-clock                #    1.000 CPUs utilized            ( +-  0.04% )
             88      context-switches          #    0.001 K/sec                    ( +- 19.31% )
              0      cpu-migrations            #    0.000 K/sec
          1,725      page-faults               #    0.022 K/sec
268,707,676,391      cycles                    #    3.498 GHz                      ( +-  0.04% )
256,391,503,899      instructions              #    0.95  insn per cycle           ( +-  0.00% )
 12,503,667,389      branches                  #  162.771 M/sec                    ( +-  0.00% )
    290,932,794      branch-misses             #    2.33% of all branches          ( +-  0.39% )

        76.8217 +- 0.0337 seconds time elapsed  ( +-  0.04% )
```

@wsnyder: With that, I think this can now be merged after a round of code review.

wsnyder commented on src/V3EmitCBase.h:
```diff
@@ -105,6 +127,49 @@ class EmitCBaseVisitor VL_NOT_FINAL : public AstNVisitor {
+    static bool isWordChar(char c) { return isalnum(c) || c == '_'; }
+
+public:
+    static string replaceWord(string str, const string& from, const string& to) {
+        const size_t len = from.size();
+        UASSERT_STATIC(len > 0, "Cannot replace empty string");
+        for (size_t pos = 0; (pos = str.find(from, pos)) != string::npos; pos += len) {
+            // Only replace whole words
+            if (((pos > 0) && isWordChar(str[pos - 1])) ||  //
+                ((pos + len < str.size()) && isWordChar(str[pos + len]))) {
+                continue;
+            }
+            str.replace(pos, len, to);
+        }
+        return str;
+    }
```
Perhaps better in V3String.  Also if len is zero one would expect it to just return "" as there's no replacements to do.

gezalore replied:
Moved to V3String. The assert is over the length of the substring to replace, i.e.: this blows up: `replaceWord("banana", "", "HUH??")`

wsnyder commented on src/V3EmitC.cpp:
```diff
@@ -226,67 +230,60 @@ class EmitCStmts VL_NOT_FINAL : public EmitCBaseVisitor {
-                    && !funcp->dpiImport()) {  // DPI is prototyped in __Dpi.h
-                    funcsp.push_back(funcp);
-                }
+                if (funcp->dpiImport())  // DPI import functions are declared in __Dpi.h
+                    continue;
+                if (funcp->isMethod() != inClassBody)  // Only methods go inside class
+                    continue;
+                if (funcp->isMethod() && funcp->isLoose())  // Loose methods are declared lazily
+                    continue;
```
Think it was easier to read as an if with several conditions but ok either way.

wsnyder commented on src/V3EmitC.cpp:
```diff
@@ -1337,16 +1377,85 @@ class EmitVarTspSorter final : public V3TSP::TspStateBase {
+//######################################################################
+// Emit lazy forward declarations
+
+class LazyDecls final : public AstNVisitor {
```
```suggestion
class EmitLazyDecls final : public AstNVisitor {
```

wsnyder commented on src/V3EmitC.cpp:
```diff
@@ -1337,16 +1377,85 @@ class EmitVarTspSorter final : public V3TSP::TspStateBase {
+    // MEMBERS
+    std::unordered_set<string> m_emittedManually;  // Set of names already declared manually.
+    EmitCBaseVisitor& m_emitter;  // For access to file output
+    bool m_needsBlankLine;  // Emit blank line if any declarations are emitted (purely cosmetic)
```
```suggestion
    bool m_needsBlankLine = false;  // Emit blank line if any declarations are emitted (purely cosmetic)
```

gezalore replied:
This would be redudant, as it's reset in the `emit` function below where this member is read. Could rewite that to reset to false on exit rather than entry if you insist.

wsnyder replied:
Any variable not reset here or in the constructor will cause some lint tools to complain.  I see your point about needing it in emit, but see no harm in also defaulting it, just good defense.

gezalore replied:
Ok, lint tools are fair point. I think I have spent too much time trying to strip redundancy from the generated code to be comfortable with these kind of things ;)

wsnyder commented on src/V3EmitC.cpp:
```diff
@@ -1337,16 +1377,85 @@ class EmitVarTspSorter final : public V3TSP::TspStateBase {
+
+    virtual void visit(AstNodeCCall* nodep) override {
+        lazyDeclare(nodep->funcp());
+        // Assume args don't contain calls themselves, which is true for now, so don't iterate ...
```
Is there a downside of iterating?   Generally I try to iterate as later when we change the visitor to do something more not iterating may be a source of bugs.

gezalore replied:
Good point, there is no downside so fixed.

wsnyder commented on src/V3EmitC.cpp:
```diff
@@ -1337,16 +1377,85 @@ class EmitVarTspSorter final : public V3TSP::TspStateBase {
+    LazyDecls(EmitCBaseVisitor& emitter)
+        : m_emitter(emitter) {}
+    void emit(AstNode* nodep) {
+        m_needsBlankLine = false;
```
```suggestion
```

gezalore replied:
Again, this is inside the `emit` function, not the constructor, basically we want a blank line every time `emit` is invoked and actually emits a declaration.

wsnyder commented on src/V3EmitC.cpp:
```diff
@@ -2775,30 +2905,56 @@ void EmitCImp::emitSettleLoop(const std::string& eval_call, bool initial) {
-    puts(/**/ "return __VlSymsp->_vm_contextp__;\n");
+void EmitCImp::emitWrapFast() {
+    UASSERT_OBJ(m_modp->isTop(), m_modp, "Attempting to emitWrapFast for non-top class");
+    puts("\nVerilatedContext* " + topClassName() + "::contextp() {\n");
```
```suggestion
    puts("\nVerilatedContext* " + topClassName() + "::contextp() const {\n");
```

Maybe.

wsnyder commented on src/V3EmitC.cpp:
```diff
@@ -3901,6 +4060,7 @@ class EmitCTrace final : EmitCStmts {
                 putsDecoration("// Final\n");
                 iterateAndNextNull(nodep->finalsp());
             }
+            m_useSelfForThis = false;
```
Remove and use a VL_RESTORER earlier instead

gezalore replied:
Fancy writing nested function definitions? :) Ok, will do.

wsnyder reviewed (commented):
Mostly minor stuff.  Is it true that split-cfuncs is basically unaffected?

I would like to push a release before we merge this into master.  I'm working through some minor issues to get that ready in the next few days.

gezalore commented:
> Is it true that split-cfuncs is basically unaffected?

Yes, the output code is basically the same and comes in the same files, you can `icdiff --recursive` the output code before and after this and it should all match up bar lazy declarations and signature changes.

> I would like to push a release before we merge this into master. I'm working through some minor issues to get that ready in the next few days.

Sounds good to me.

gezalore commented on src/V3File.cpp:
```diff
@@ -965,6 +965,7 @@ class VIdProtectImp final {
 public:
     VIdProtectImp() {
         passthru("this");
+        passthru("self");
```
Meh, probably, that will only fail on a user having a `vlSelf` I presume, which I accept is statistically better.

gezalore replied:
Actually, wherever the internal self pointer is present, every variable acces should go through an explicit dereference, and we removed the #define self this, so this should be unambiguous shouldn't it? Worst that can happen is you see `self->self` in the code, but not many people read that.

wsnyder replied:
You might be right, but I can't intuit if anything breaks. Perhaps add a test that has a DPI import and export named self?, and it passes, then I think we're good.

gezalore replied:
Hmm, ok, a DPI import of a function called `self` would fail as the local name would shadow the global function, so will go with `vlSelf`.

gezalore commented:
Ok, I believe It's a go after the release unless you have further thoughts.

wsnyder commented:
Feel free to squash push this on 6/13 or later when you feel it's stable.

wsnyder commented:
Also please update "Changes" to describe this - a good general practice too (but can cause some merge problems unfortunately so better to wait until nearly ready to merge)
